### PR TITLE
Adding Container Exec Command for Azure Container Instance

### PR DESCRIPTION
**Pull request by samkreter**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](../doc/command_guidelines.md).

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/5854
This is an **experimental preview** for @microsoft users.

**Comment by tjprescott** [comment]
Please show the output of this command's help with "-h"

**Comment by samkreter** [comment]
image

**Comment by tjprescott** [comment]
The help text for `--exec-command` doesn't make sense to me.  Is it not the command to be executed?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_help.py [hunk 1]
```diff
@@ -57,6 +57,11 @@
     short-summary: Examine the logs for a container in a container group.
 """
 
+helps['container exec'] = """
+    type: command
+    short-summary: Exec command for a container in a container group.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/_help.py]: "Execute command..."?  

Also, this is a non-trivial command and thus should have an example.  If this is only supported on Linux, then it should be called out in a long-summary.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py [hunk 2]
```diff
@@ -71,5 +71,11 @@ def load_arguments(self, _):
         c.argument('container_name', help='The container name to tail the logs. If omitted, the first container in the container group will be chosen')
         c.argument('follow', help='Indicate to stream the tailing logs', action='store_true')
 
+    with self.argument_context('container exec') as c:
+        c.argument('container_name', help='The container name to exec the command. If omitted, the first container in the container group will be chosen')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py]: Regarding the default, is this an arbitrary choice?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py [hunk 3]
```diff
@@ -71,5 +71,11 @@ def load_arguments(self, _):
...
+        c.argument('exec_command', help='The container to exec to.')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py]: Is this help text correct?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py [hunk 4]
```diff
@@ -71,5 +71,11 @@ def load_arguments(self, _):
...
+        c.argument('terminal_row_size', help='The size for the terminal rows.')
+        c.argument('terminal_col_size', help='The size for the terminal columns.')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py]: What are these for?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/commands.py [hunk 5]
```diff
@@ -16,4 +16,5 @@ def load_command_table(self, _):
                          table_transformer=transform_container_group)
         g.custom_command('delete', 'delete_container', confirmation=True)
         g.custom_command('logs', 'container_logs', client_factory=cf_container_logs)
+        g.custom_command('exec', 'container_exec')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/commands.py]: This should be a verb, not a platform-specific command name. I would suggest something like "run" or "run-command" (which has precedent in the VM module).

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 6]
```diff
@@ -114,7 +119,8 @@ def create_container(client,
                             image_registry_credentials=image_registry_credentials,
                             volumes=volumes or None)
 
-    return client.create_or_update(resource_group_name, name, cgroup)
+    raw_response = client.create_or_update(resource_group_name, name, cgroup, raw=True)
+    return raw_response.output
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: What is the reason for this change?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 7]
```diff
@@ -231,6 +237,63 @@ def container_logs(cmd, resource_group_name, name, container_name=None, follow=F
             stream_args=(logs_client, resource_group_name, name, container_name, container_group.restart_policy))
 
 
+def container_exec(cmd, resource_group_name, name, container_name=None, exec_command=None, terminal_row_size=None, terminal_col_size=None):
+    """Start exec for a container. """
+
+    if platform.system() is WINDOWS_OS_NAME:
+        print("The container exec operation is currently not supported for Windows.")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md
"Log to logger.error() or logger.warning() for user messages; do not use the print() function"

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 8]
```diff
@@ -231,6 +237,63 @@ def container_logs(cmd, resource_group_name, name, container_name=None, follow=F
...
+    start_container_client = cf_start_container(cmd.cli_ctx)
+    container_group_client = cf_container_groups(cmd.cli_ctx)
+    container_group = container_group_client.get(resource_group_name, name)
+
+    # Set defaults for rows and cols size
+    if terminal_row_size is None:
+        terminal_row_size = 24
+
+    if terminal_col_size is None:
+        terminal_col_size = 80
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: You can and should just put these default values in the command signature. Then they will appear when the user uses `-h`

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 9]
```diff
@@ -231,6 +237,63 @@ def container_logs(cmd, resource_group_name, name, container_name=None, follow=F
...
+
+    # If container name is not present, use the first container.
+    if container_name is None:
+        container_name = container_group.containers[0].name
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: Unless there is a compelling reason that the first container is the one the user would most like to use or the most frequently used, consider simply making this a required parameter.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 10]
```diff
@@ -231,6 +237,63 @@ def container_logs(cmd, resource_group_name, name, container_name=None, follow=F
...
+                    if not x:
+                        break
+                    ws.send(x)
+            except (select.error, IOError) as e:
+                if e.args and e.args[0] == errno.EINTR:
+                    pass
+                else:
+                    raise
+    except websocket.WebSocketException:
+        print("Connection Ended.")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md
"Log to logger.error() or logger.warning() for user messages; do not use the print() function"

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_help.py (hunk 1 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_help.py]: Also, from reviewing this PR, I don't think this help text is clear at all. I initially thought this command would execute a command on a specific container instance.  Now it seems like it basically opens a shell that tunnels into that container instance to run any number of commands.  Is this right?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py (hunk 2 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py]: Using the first container in a group is the standard since many container groups have only 1 container it will default to that.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py (hunk 4 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py]: They are the size of the output for the command. I agree they should have much better help text. I will fix that now.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/commands.py (hunk 5 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/commands.py]: Exec is the standard in the Kubernetes/Docker community.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py (hunk 6 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: We are preserving the current CLI output for create/update container group which is immediately return. Since this Sdk introduced LRO, we want to wait until the next release to have functionality change.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_help.py (hunk 1 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_help.py]: I agree.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py (hunk 8 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: Done

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py (hunk 2 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py]: That makes sense. Alternatively, you could make it the default *if* there's only one.  If there's more than one, you could either:
1) fail and require the user specify one (makes sense if this could be part of a script)
2) prompt the user to specify one (makes sense if this is part of an interactive flow)

I'm not sure which scenarios are supported here, so I can't say which would be better. However, arbitrarily choosing the first if there are more than one might surprise/upset users.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_help.py (hunk 1 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_help.py]: To address the second part. It will execute any command from within the container and stream the output back. The most common use case is to exec bash a the command which opens and interactive terminal from within the container. The example now shows doing this.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 11]
```diff
@@ -231,6 +245,73 @@ def container_logs(cmd, resource_group_name, name, container_name=None, follow=F
...
+    start_container_client = cf_start_container(cmd.cli_ctx)
+    container_group_client = cf_container_groups(cmd.cli_ctx)
+    container_group = container_group_client.get(resource_group_name, name)
+
+    if len(container_group.containers) > 1 and container_name is None:
+        logger.error("Must specify the container name for container groups with more than one container.")
+
+    # If only one container in container group, use that container.
+    if container_name is None:
+        container_name = container_group.containers[0].name
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: Should be something like:
```Python
if container_name is None and len(container_group.containers) == 1:
  # what you have
else:
  raise CLIError('--container-name required when container group has more than one container.')
```
